chainer/optimizers/ada_delta.py

```diff
@@ -4,43 +4,83 @@
-            state['msdx'] = xp.zeros_like(data)
+    Args:
+        rho (float): Exponential decay rate of the first and second order
+            moments.
+        eps (float): Small value for the numerical stability.
 
-    def update_one_cpu(self, param, state):
+    """
+    def __init__(self, rho=None, eps=None):
+        super(AdaDeltaRule, self).__init__()
+        self.hyperparam = optimizer.Hyperparameter(_default_hyperparam)
+        if rho is not None:
+            self.hyperparam.rho = rho
+        if eps is not None:
+            self.hyperparam.eps = eps
+
+    def init_state(self, param):
+        xp = cuda.get_array_module(param.data)
+        with cuda.get_device(param.data):
+            self.state['msg'] = xp.zeros_like(param.data)
+            self.state['msdx'] = xp.zeros_like(param.data)
+
+    def update_core_cpu(self, param):
         grad = param.grad
-        msg, msdx = state['msg'], state['msdx']
+        msg, msdx = self.state['msg'], self.state['msdx']
+        rho = self.hyperparam.rho
+        eps = self.hyperparam.eps
 
-        msg *= self.rho
-        msg += (1 - self.rho) * grad * grad
-        dx = numpy.sqrt((msdx + self.eps) / (msg + self.eps)) * grad
-        msdx *= self.rho
-        msdx += (1 - self.rho) * dx * dx
+        msg *= rho
+        msg += (1 - rho) * grad * grad
+        dx = numpy.sqrt((msdx + eps) / (msg + eps)) * grad
+        msdx *= rho
+        msdx += (1 - rho) * dx * dx
         param.data -= dx
 
-    def update_one_gpu(self, param, state):
+    def update_core_gpu(self, param):
         cuda.elementwise(
             'T grad, T one_minus_rho, T eps',
             'T param, T msg, T msdx',
             '''msg   = msg + one_minus_rho * (grad * grad - msg);
                T dx  = sqrt((msdx + eps) / (msg + eps)) * grad;
                msdx  += one_minus_rho * (dx * dx - msdx);
                param -= dx;''',
-            'adadelta')(param.grad, 1 - self.rho, self.eps,
-                        param.data, state['msg'], state['msdx'])
+            'adadelta')(param.grad, 1 - self.hyperparam.rho,
+                        self.hyperparam.eps, param.data,
+                        self.state['msg'], self.state['msdx'])
+
+
+class AdaDelta(optimizer.GradientMethod):
+
+    """Zeiler's ADADELTA.
+
+    See: http://www.matthewzeiler.com/pubs/googleTR2012/googleTR2012.pdf
+
+    Args:
+        rho (float): Exponential decay rate of the first and second order
+            moments.
+        eps (float): Small value for the numerical stability.
+
+    """
+    def __init__(self, rho=_default_hyperparam.rho,
+                 eps=_default_hyperparam.eps):
+        super(AdaDelta, self).__init__()
+        self.hyperparam = optimizer.Hyperparameter()
+        self.hyperparam.rho = rho
+        self.hyperparam.eps = eps
+
+    def setup_update_rule(self, param):
+        param.update_rule = AdaDeltaRule()
+        param.update_rule.hyperparam.set_parent(self.hyperparam)
```

Comment:
It seems what optimizers do in `setup_update_rule` is same (i.e. it adds an instance of `UpdateRule` to the parameter  and set an instance of `HyperParameter` which acts as default hyper parameters). The difference is the type of `UpdateRule` only. Do you expect the developer of custom optimizers would like to do other than the typical procedure? If not, I think just providing an instance of UpdateRule is simpler and not confusing.